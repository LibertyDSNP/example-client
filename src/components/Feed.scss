.Feed__block {
  background-color: $gray;
  padding: 16px;
  display: flex;
  flex-direction: column;
  flex-basis: 50%;
}

.Feed__header {
  display: flex;
  justify-content: space-between;
}
