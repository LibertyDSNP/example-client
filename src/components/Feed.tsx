import React from "react";
import NewPost from "./NewPost";
import PostList from "./PostList";

const Feed: React.FC = () => {
  return (
    <div className="Feed__block">
      <div className="Feed__header">
        <h1>Feed</h1>
        <NewPost />
      </div>
      <PostList />
    </div>
  );
};
export default Feed;
