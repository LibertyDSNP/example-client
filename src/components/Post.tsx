import React from "react";
import { Card } from "antd";
import { FeedItem, Profile } from "../utilities/types";
import UserAvatar from "./UserAvatar";
import PostMedia from "./PostMedia";
import { ActivityContentAttachment } from "@dsnp/sdk/core/activityContent";
import ActionsBar from "./ActionsBar";
import { DSNPUserId } from "@dsnp/sdk/dist/types/core/identifiers";
import { useAppSelector } from "../redux/hooks";

interface PostProps {
  feedItem: FeedItem;
}

const Post = ({ feedItem }: PostProps): JSX.Element => {
  const noteContent = feedItem.content;
  const attachments = noteContent.attachment;

  const profiles: Record<DSNPUserId, Profile> = useAppSelector(
    (state) => state.profiles?.profiles || {}
  );
  return (
    <Card key={feedItem.hash} className="Post__block" bordered={false}>
      <Card.Meta
        className="Post__header"
        avatar={
          <UserAvatar profileAddress={feedItem.fromId} avatarSize={"medium"} />
        }
        title={profiles[feedItem.fromId].name || feedItem.fromId}
        description={
          <div className="Post__description">
            @{profiles[feedItem.fromId].handle}
          </div>
        }
      />
      <PostMedia attachment={attachments as ActivityContentAttachment[]} />
      <div className="Post__caption">
        <ActionsBar published={feedItem.published} />
        <div>{noteContent.content}</div>
        <div className="Post__captionTags">
          {feedItem.tags && feedItem.tags.join(" ")}
        </div>
      </div>
    </Card>
  );
};

export default Post;
