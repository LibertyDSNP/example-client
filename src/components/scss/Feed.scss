.Feed__block {
  background-color: $white;
  border-radius: 25px;
  padding: 16px;
  display: flex;
  flex-direction: column;
  width: 750px;
  height: fit-content;
  margin-right: 15px;
}

.Feed__header {
  display: flex;
  justify-content: space-between;
  border-bottom: 2px solid $black;
  padding-bottom: 12px;
}

.Feed__navigation {
  display: flex;
  align-items: center;
  font-weight: bold;
}

.Feed__navigationItem {
  padding: 0 10px;
  cursor: pointer;
}

.Feed__navigationItem:first-child {
  padding-left: 0;
}

.Feed__navigationItem--active {
  text-decoration: underline;
}

.Feed__navigationSpacer {
  width: 5px;
  height: 22px;
  transform: skew(-20deg);
  background: $green;
}
