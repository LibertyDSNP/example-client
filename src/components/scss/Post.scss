.Post__block {
  width: 50vw;
  position: relative;
}

.Post__caption {
  margin: 16px 0;
}
