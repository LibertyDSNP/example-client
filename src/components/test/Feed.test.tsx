import Feed from "../Feed";
import Post from "../Post";
import { mount, shallow } from "enzyme";
import { getPrefabFeed } from "../../test/testFeeds";
import { componentWithStore, createMockStore } from "../../test/testhelpers";
import { getPreFabSocialGraph } from "../../test/testGraphs";
import { getPrefabProfile } from "../../test/testProfiles";

const userId = getPrefabProfile(0).fromId;
const feedItems = getPrefabFeed();
const graphs = getPreFabSocialGraph();
const initialState = {
  user: { id: userId },
  feed: {
    feedItems: feedItems,
    isPostLoading: { loading: false, myIdentifier: undefined },
    isReplyLoading: { loading: false, parent: undefined },
  },
  graphs: graphs,
};
const store = createMockStore(initialState);

describe("Feed", () => {
  it("renders without crashing", () => {
    expect(() => {
      shallow(componentWithStore(Feed, store));
    }).not.toThrow();
  });

  it("does display new post button when logged in", () => {
    const component = mount(componentWithStore(Feed, store));
    expect(component.find(".Feed__newPostButton").length).not.toBe(0);
  });

  it("does not display new post button when not logged in", () => {
    const initialState = {
      user: {},
      feed: {
        feedItems,
        isPostLoading: { loading: false, myIdentifier: undefined },
        isReplyLoading: { loading: false, parent: undefined },
      },
      graphs: { graphs },
    };
    const store = createMockStore(initialState);
    const component = mount(componentWithStore(Feed, store));
    expect(component.find(".Feed__newPostButton").length).toBe(0);
  });

  it("opens new post modal onclick", () => {
    const component = mount(componentWithStore(Feed, store));
    component.find(".Feed__newPostButton").first().simulate("click");
    expect(component.find("Modal")).toBeTruthy();
  });

  describe("Displays Correct Feed", () => {
    it("Connections Feed", () => {
      const component = mount(componentWithStore(Feed, store));
      expect(component.find(Post).length).toEqual(3);

      const expectedFeedAddresses = [userId].concat(
        Object.keys(graphs.following)
      );
      component.find(".ant-card-meta-title").forEach((address) => {
        expect(expectedFeedAddresses).toContain(address.text());
      });
    });

    it("My Feed", () => {
      const component = mount(componentWithStore(Feed, store));
      const button = component.findWhere((node) => {
        return (
          node.hasClass("Feed__navigationItem") && node.text() === "My Posts"
        );
      });
      button.simulate("click");
      expect(component.find(Post).length).toEqual(2);
      component.find(".ant-card-meta-title").forEach((address) => {
        expect(address.text()).toBe(userId);
      });
    });

    it("All Posts", () => {
      const component = mount(componentWithStore(Feed, store));
      const button = component.findWhere((node) => {
        return (
          node.hasClass("Feed__navigationItem") && node.text() === "All Posts"
        );
      });
      button.simulate("click");
      expect(component.find(Post).length).toEqual(4);
    });
  });
});
