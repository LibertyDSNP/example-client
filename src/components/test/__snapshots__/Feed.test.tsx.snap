// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed does display new post button when logged in 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Feed>
        <div
          className="Feed__block"
        >
          <div
            className="Feed__header"
          >
            <h1>
              Feed
            </h1>
            <Button
              className="Feed__newPostButton"
              onClick={[Function]}
            >
              <Wave>
                <button
                  className="ant-btn Feed__newPostButton"
                  onClick={[Function]}
                  type="button"
                >
                  <LoadingIcon
                    existIcon={false}
                    loading={false}
                    prefixCls="ant-btn"
                  >
                    <CSSMotion
                      motionName="ant-btn-loading-icon-motion"
                      onAppearActive={[Function]}
                      onAppearStart={[Function]}
                      onEnterActive={[Function]}
                      onEnterStart={[Function]}
                      onLeaveActive={[Function]}
                      onLeaveStart={[Function]}
                      removeOnLeave={true}
                      visible={false}
                    >
                      <DomWrapper />
                    </CSSMotion>
                  </LoadingIcon>
                  <span
                    key=".0"
                  >
                    New Post
                  </span>
                </button>
              </Wave>
            </Button>
          </div>
          <PostList>
            <div
              className="PostList__block"
            >
              <h1>
                PostList
              </h1>
              <div
                key="0"
              >
                <div>
                  0x1DCODE3333
                </div>
                <div>
                  My favorite dessert is Cake
                </div>
                <img
                  alt=""
                  width={400}
                />
              </div>
              <div
                key="1"
              >
                <div>
                  0x1DCODE0000
                </div>
                <div>
                  Going to the mall!
                </div>
                <img
                  alt=""
                  width={400}
                />
              </div>
              <div
                key="2"
              >
                <div>
                  0x1DCODE0000
                </div>
                <div>
                  Hello World
                </div>
                <img
                  alt=""
                  width={400}
                />
              </div>
              <div
                key="3"
              >
                <div>
                  0x1DCODE2222
                </div>
                <div>
                  Everyone leave me alone
                </div>
                <img
                  alt="Image for post"
                  src="https://64.media.tumblr.com/bd8d2127a91f57463c2e753cf837ab6e/014df86f4004efef-ec/s1280x1920/adda1023806b71606f83f484a64daa03bce12c8d.jpg"
                  width={400}
                />
              </div>
            </div>
          </PostList>
        </div>
      </Feed>
    </Router>
  </BrowserRouter>
</Provider>
`;
